File: 122-best-time-to-buy-and-sell-stock-ii/best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int solve(int i, int[] prices, int n, int buyFlag, int[][] dp){
        if(i == n){
            return 0;
        }

        if(dp[i][buyFlag] != -1){
            return dp[i][buyFlag];
        }

        if(buyFlag == 1){
            int buy = solve(i + 1, prices, n, 0, dp) - prices[i];
            int notBuy = solve(i + 1, prices, n, 1, dp);

            return dp[i][buyFlag] = Math.max(buy, notBuy);
        }
        else{
            int sell = solve(i + 1, prices, n, 1, dp) + prices[i];
            int notSell = solve(i + 1, prices, n, 0, dp);

            return dp[i][buyFlag] = Math.max(sell, notSell);
        }
    }
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int dp[][] = new int[n + 1][2];
        for(int i = 0;  i <= n; i++){
            for(int j = 0; j < 2; j++){
                dp[i][j] = -1;
            }
        }
        return solve(0, prices, n, 1, dp);
    }
}

File: 122-best-time-to-buy-and-sell-stock-ii/best-time-to-buy-and-sell-stock-ii.py
class Solution(object):
    def solve(self, i, prices, buy_flag, dp):
        if i == len(prices):
            return 0
        if dp[i][buy_flag] != None:
            return dp[i][buy_flag]
        if buy_flag:
            buy = self.solve(i + 1, prices, 0, dp) - prices[i]
            not_buy = self.solve(i + 1, prices, 1, dp)
            dp[i][buy_flag] = max(buy, not_buy)
            return dp[i][buy_flag]
        else:
            sell = self.solve(i + 1, prices, 1, dp) + prices[i]
            not_sell = self.solve(i + 1, prices, 0, dp)
            dp[i][buy_flag] = max(sell, not_sell)
            return dp[i][buy_flag]

    def maxProfit(self, prices):
        dp = [[None] * 2 for _ in range(len(prices))] 
        return self.solve(0, prices, 1, dp)
        